java/src/main/native/src/row_conversion.cu

```diff
@@ -447,32 +1073,558 @@ static inline int32_t compute_fixed_width_layout(std::vector<cudf::data_type> co
+    bool nested_type = col_type.id() == type_id::LIST || col_type.id() == type_id::STRING;
+
+    // a list or string column will write a single uint64
+    // of data here for offset/length
+    auto col_size = nested_type ? 8 : size_of(col_type);
+
+    // align size for this type
+    std::size_t const alignment_needed = col_size; // They are the same for fixed width types
+    fixed_width_size_per_row = detail::align_offset(fixed_width_size_per_row, alignment_needed);
+    column_starts.push_back(fixed_width_size_per_row);
+    column_sizes.push_back(col_size);
+    fixed_width_size_per_row += col_size;
+  }
+
+  auto validity_offset = fixed_width_size_per_row;
+  column_starts.push_back(validity_offset);
+
+  return fixed_width_size_per_row;
+}
+
+/**
+ * @brief Build `block_info` for the validity data to break up the work.
+ *
+ * @param num_columns number of columns in the table
+ * @param num_rows number of rows in the table
+ * @param shmem_limit_per_block size of shared memory available to a single gpu block
+ * @param row_batches batched row information for multiple output locations
+ * @return vector of `block_info` structs for validity data
+ */
+std::vector<detail::block_info>
+build_validity_block_infos(size_type const &num_columns, size_type const &num_rows,
+                           size_type const &shmem_limit_per_block,
+                           std::vector<row_batch> const &row_batches) {
+  auto const desired_rows_and_columns = (int)sqrt(shmem_limit_per_block);
+  auto const column_stride = align_offset(
+      [&]() {
+        if (desired_rows_and_columns > num_columns) {
+          // not many columns, group it into 8s and ship it off
+          return std::min(8, num_columns);
+        } else {
+          return util::round_down_safe(desired_rows_and_columns, 8);
+        }
+      }(),
+      8);
+  // we fit as much as we can given the column stride
+  // note that an element in the table takes just 1 bit, but a row with a single
+  // element still takes 8 bytes!
+  auto const bytes_per_row = align_offset(util::div_rounding_up_unsafe(column_stride, 8), 8);
+  auto const row_stride = std::min(num_rows, shmem_limit_per_block / bytes_per_row);
+
+  std::vector<detail::block_info> validity_block_infos;
+  for (int col = 0; col < num_columns; col += column_stride) {
+    int current_window_row_batch = 0;
+    int rows_left_in_batch = row_batches[current_window_row_batch].row_count;
+    int row = 0;
+    while (row < num_rows) {
+      if (rows_left_in_batch == 0) {
+        current_window_row_batch++;
+        rows_left_in_batch = row_batches[current_window_row_batch].row_count;
+      }
+      int const window_height = std::min(row_stride, rows_left_in_batch);
+
+      validity_block_infos.emplace_back(detail::block_info{
+          col, row, std::min(col + column_stride - 1, num_columns - 1), row + window_height - 1});
+      row += window_height;
+      rows_left_in_batch -= window_height;
```

Comment:
How big is window_height here compared to # of rows?  Wondering about many times we could end up looping here.

Reply:
This is typically 144 bytes across and 144 rows. This doesn't always work out due to edges and the size of the columns, but a full window is half the shared memory, which works out to 144 bytes/rows.

Reply:
So the # of iterations is something like (# rows) / 144?

Reply:
Seems like this whole thing can be done in parallel.  It's returning host data that gets immediately copied to the device anyway.  Maybe something like (roughly):

```
auto size_type num_row_windows = reduction over row_batches[].row_count;
auto const num_col_windows = num_columns / column_stride;
auto const total_infos = num_row_windows * num_col_windows;
rmm::device_uvector<detail::block_info> validity_block_infos(total_infos);
auto iter = thrust::make_counting_iterator(0);
thrust::transform(..., iter, iter + total_infos, validity_block_infos.begin(), []__device__(size_type i){
   size_type const col_batch_index = i / num_row_windows;
   size_type const col = col_batch_index * column_stride;
   size_type const row_batch_index = i % num_row_windows;
   size_type const row = row_batch_index * window_height;
   
   // seems like we should be able to derive everything we need at this point
   return detail::block_info{...}
});
```

Reply:
Thinking about this some more: do we even need to do this?  Do the kernels run in a way such that we can just compute this info on the spot?

Reply:
We could get the validity kernels running based on knowing the width and height value, but not the data kernels since the column widths are different. The potential gotcha is figuring out which batch the block uses. That's a good suggestion and something I should investigate, but probably will turn into a fairly complex task given the batch issue. Should this be an issue to address later?